flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataManager.java

```diff
@@ -0,0 +1,673 @@
+                                                        ConsumeStatus.NOT_CONSUMED)) {
+                                                    checkNotNull(
+                                                                    bufferContext.spilledFuture,
+                                                                    "Buffer in spill status should already set spill future.")
+                                                            .thenRun(
+                                                                    () ->
+                                                                            dataIndex
+                                                                                    .markBufferReadable(
+                                                                                            subpartitionId,
+                                                                                            bufferIndex));
+                                                }
+                                                subpartitionDataBuffer.onBufferReleased(
+                                                        bufferIndex);
+                                            });
+                                }
+                            });
+                });
+    }
+
+    private void recycleBuffer(MemorySegment buffer) {
+        numRequestedBuffers.decrementAndGet();
+        bufferPool.recycle(buffer);
+    }
+
+    class HsSubpartitionDataBuffer {
+        // this field only accessed by task main thread.
+        private final Queue<BufferBuilder> unfinishedBuffers = new LinkedList<>();
+
+        // this field only accessed by task main thread.
+        private int finishedBufferIndex;
+
+        /**
+         * Lock for subpartition's own field. All buffers status belong to this subpartition guarded
+         * by this lock.
+         */
+        private final Object subpartitionLock = new Object();
+
+        // TODO maybe we should rename this to allBuffers or unReleasedBuffers?
+        @GuardedBy("subpartitionLock")
+        private final Deque<BufferContext> finishedBuffers = new LinkedList<>();
+
+        @GuardedBy("subpartitionLock")
+        private final Deque<BufferContext> consumeBuffers = new LinkedList<>();
+
+        @GuardedBy("subpartitionLock")
+        private final Map<Integer, BufferContext> bufferIndexToContexts;
+
+        private final int targetChannel;
+
+        HsSubpartitionDataBuffer(int targetChannel) {
+            this.targetChannel = targetChannel;
+            this.bufferIndexToContexts = new HashMap<>();
+        }
+
+        // ------------------------------------------------------------------------
+        //  Called by Consumer
+        // ------------------------------------------------------------------------
+
+        public Buffer.DataType getFirstDataType(int expectedBufferIndex) {
+            synchronized (subpartitionLock) {
```

Comment:
Why not guarded by the read lock?

Reply:
I'd suggest to abstract the lock accessing logics with methods like: `runWithSubpartitionLock` and `runWithPartitionLock`

Reply:
done.